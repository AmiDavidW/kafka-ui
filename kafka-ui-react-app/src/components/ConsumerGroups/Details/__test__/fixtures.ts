export const clusterName = 'testClusterName';

export const name = 'testName';
export const partition = {
  topic: 'someTopicName',
  partition: 0,
  currentOffset: 2597,
  endOffset: 2598,
  messagesBehind: 1,
  consumerId: '_someConsumerId',
  host: '/0.0.0.0',
};
