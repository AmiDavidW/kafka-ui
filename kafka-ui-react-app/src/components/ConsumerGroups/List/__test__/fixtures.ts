import { ConsumerGroupState } from 'generated-sources';

export const consumerGroup1 = {
  groupId: 'group1',
  members: 1,
  topics: 1,
  simple: false,
  partitionAssignor: 'range',
  state: ConsumerGroupState.STABLE,
  coordinator: {
    id: 2,
    host: 'dev-cp-kafka-0.dev-cp-kafka-headless.kafka-ui',
    port: 9092,
  },
  messagesBehind: 0,
};

export const consumerGroup2 = {
  groupId: 'group2',
  members: 1,
  topics: 1,
  simple: false,
  partitionAssignor: 'range',
  state: ConsumerGroupState.STABLE,
  coordinator: {
    id: 2,
    host: 'dev-cp-kafka-0.dev-cp-kafka-headless.kafka-ui',
    port: 9092,
  },
  messagesBehind: 0,
};

export const consumerGroup = consumerGroup1;

export const noConsumerGroupsResponse = {
  pageCount: 1,
  consumerGroups: [],
};
export const someComnsumerGroupsResponse = {
  pageCount: 1,
  consumerGroups: [consumerGroup1, consumerGroup2],
};

export const searchComnsumerGroupsResponse = {
  pageCount: 1,
  consumerGroups: [consumerGroup1],
};
