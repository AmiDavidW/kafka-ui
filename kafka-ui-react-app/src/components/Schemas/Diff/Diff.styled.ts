import styled from 'styled-components';
import { Colors } from 'theme/theme';

export const DiffWrapper = styled.div`
  align-items: stretch;
  display: block;
  flex-basis: 0;
  flex-grow: 1;
  flex-shrink: 1;
  min-height: min-content;
  padding-top: 1.5rem !important;
  background: blue;
  &
    .ace_editor
    > .ace_scroller
    > .ace_content
    > .ace_marker-layer
    > .codeMarker {
    background: ${Colors.yellow[20]};
    position: absolute;
    z-index: 20;
  }
`;
