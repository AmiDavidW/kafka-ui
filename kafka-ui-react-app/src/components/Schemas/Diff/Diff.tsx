import React from 'react';
import { SchemaSubject } from 'generated-sources';
import { ClusterName, SchemaName } from 'redux/interfaces';
import { clusterSchemaSchemaDiffPath } from 'lib/paths';
import PageLoader from 'components/common/PageLoader/PageLoader';
import DiffViewer from 'components/common/DiffViewer/DiffViewer';
import { useHistory } from 'react-router';
import { fetchSchemaVersions } from 'redux/reducers/schemas/schemasSlice';

import * as S from './Diff.styled';

export interface DiffProps {
  subject: SchemaName;
  clusterName: ClusterName;
  leftVersionInPath?: string;
  rightVersionInPath?: string;
  versions: SchemaSubject[];
  areVersionsFetched: boolean;
}

const Diff: React.FC<DiffProps> = ({
  subject,
  clusterName,
  leftVersionInPath,
  rightVersionInPath,
  versions,
  areVersionsFetched,
}) => {
  const [leftVersion, setLeftVersion] = React.useState(leftVersionInPath || '');
  const [rightVersion, setRightVersion] = React.useState(
    rightVersionInPath || ''
  );

  React.useEffect(() => {
    fetchSchemaVersions({ clusterName, subject });
  }, [fetchSchemaVersions, clusterName]);

  const getSchemaContent = (allVersions: SchemaSubject[], version: string) => {
    const selectedSchema =
      allVersions.find((s) => s.version === version)?.schema ||
      (allVersions.length ? allVersions[0].schema : '');
    return selectedSchema.trim().startsWith('{')
      ? JSON.stringify(JSON.parse(selectedSchema), null, '\t')
      : selectedSchema;
  };
  const history = useHistory();

  return (
    <div className="section">
      {areVersionsFetched ? (
        <div className="box tile is-ancestor is-vertical">
          <div className="tile">
            <div className="tile is-6">
              <div className="select">
                <select
                  id="left-select"
                  defaultValue={leftVersion}
                  onChange={(event) => {
                    history.push(
                      clusterSchemaSchemaDiffPath(
                        clusterName,
                        subject,
                        event.target.value,
                        !rightVersion && versions.length
                          ? versions[0].version
                          : rightVersion
                      )
                    );
                    setLeftVersion(event.target.value);
                  }}
                >
                  {versions.map((version) => (
                    <option
                      key={`left-${version.version}`}
                      value={version.version}
                    >
                      {`Version ${version.version}`}
                    </option>
                  ))}
                </select>
              </div>
            </div>
            <div className="tile is-6">
              <div className="select">
                <select
                  id="right-select"
                  defaultValue={rightVersion}
                  onChange={(event) => {
                    history.push(
                      clusterSchemaSchemaDiffPath(
                        clusterName,
                        subject,
                        !leftVersion && versions.length
                          ? versions[0].version
                          : leftVersion,
                        event.target.value
                      )
                    );
                    setRightVersion(event.target.value);
                  }}
                >
                  {versions.map((version) => (
                    <option
                      key={`right-${version.version}`}
                      value={version.version}
                    >
                      {`Version ${version.version}`}
                    </option>
                  ))}
                </select>
              </div>
            </div>
          </div>
          <S.DiffWrapper>
            <DiffViewer
              value={[
                getSchemaContent(versions, leftVersion),
                getSchemaContent(versions, rightVersion),
              ]}
              setOptions={{
                autoScrollEditorIntoView: true,
              }}
              isFixedHeight={false}
            />
          </S.DiffWrapper>
        </div>
      ) : (
        <PageLoader />
      )}
    </div>
  );
};

export default Diff;
