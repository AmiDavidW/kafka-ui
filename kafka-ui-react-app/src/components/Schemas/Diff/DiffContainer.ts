import { connect } from 'react-redux';
import { ClusterName, RootState } from 'redux/interfaces';
import { RouteComponentProps, withRouter } from 'react-router-dom';
import {
  getAreSchemaVersionsFulfilled,
  selectAllSchemaVersions,
} from 'redux/reducers/schemas/schemasSlice';

import Diff from './Diff';

interface RouteProps {
  clusterName: ClusterName;
  subject: string;
  leftVersion?: string;
  rightVersion?: string;
}

type OwnProps = RouteComponentProps<RouteProps>;

const mapStateToProps = (
  state: RootState,
  {
    match: {
      params: { clusterName, subject, leftVersion, rightVersion },
    },
  }: OwnProps
) => ({
  subject,
  versions: selectAllSchemaVersions(state),
  areVersionsFetched: getAreSchemaVersionsFulfilled(state),
  clusterName,
  leftVersionInPath: leftVersion,
  rightVersionInPath: rightVersion,
});

export default withRouter(connect(mapStateToProps)(Diff));
