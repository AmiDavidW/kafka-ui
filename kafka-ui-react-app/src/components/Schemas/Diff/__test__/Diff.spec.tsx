import React from 'react';
import { Provider } from 'react-redux';
import { shallow, mount } from 'enzyme';
import configureStore from 'redux-mock-store';
import { StaticRouter } from 'react-router';
import DiffContainer from 'components/Schemas/Diff/DiffContainer';
import Diff, { DiffProps } from 'components/Schemas/Diff/Diff';

import { versions } from './fixtures';

const clusterName = 'testCluster';
const subject = 'test';
const mockStore = configureStore();

describe('Diff', () => {
  describe('Container', () => {
    const initialState: Partial<DiffProps> = {};
    const store = mockStore(initialState);

    it('renders view', () => {
      const wrapper = mount(
        <Provider store={store}>
          <StaticRouter>
            <Diff
              versions={versions}
              clusterName={clusterName}
              leftVersionInPath=""
              rightVersionInPath=""
              subject={subject}
              areVersionsFetched
            />
          </StaticRouter>
        </Provider>
      );

      expect(wrapper.exists(Diff)).toBeTruthy();
    });
  });

  describe('View', () => {
    const setupWrapper = (props: Partial<DiffProps> = {}) => (
      <Diff
        subject={subject}
        clusterName={clusterName}
        areVersionsFetched
        versions={[]}
        {...props}
      />
    );

    // describe('Initial state', () => {
    //   it('should call fetchSchemaVersions every render', () => {
    //     mount(
    //       <StaticRouter>
    //         {setupWrapper({ fetchSchemaVersions: fetchSchemaVersionsMock })}
    //       </StaticRouter>
    //     );

    //     expect(fetchSchemaVersionsMock).toHaveBeenCalledWith(
    //       clusterName,
    //       subject
    //     );
    //   });

    //   it('matches snapshot', () => {
    //     expect(
    //       shallow(
    //         setupWrapper({ fetchSchemaVersions: fetchSchemaVersionsMock })
    //       )
    //     ).toMatchSnapshot();
    //   });
    // });

    describe('when page with schema versions is loading', () => {
      const wrapper = shallow(setupWrapper({ areVersionsFetched: false }));

      it('renders PageLoader', () => {
        expect(wrapper.exists('PageLoader')).toBeTruthy();
      });

      it('matches snapshot', () => {
        expect(
          shallow(setupWrapper({ areVersionsFetched: false }))
        ).toMatchSnapshot();
      });
    });

    describe('when schema versions are loaded and no specified versions in path', () => {
      const wrapper = shallow(setupWrapper({ versions }));

      it('renders all options', () => {
        expect(wrapper.find('option')).toHaveLength(6);
      });

      it('renders left select with empty value', () => {
        expect(wrapper.find('#left-select')).toHaveLength(1);
        expect(wrapper.find('#left-select').props().defaultValue).toBe('');
      });

      it('renders right select with empty value', () => {
        expect(wrapper.find('#right-select')).toHaveLength(1);
        expect(wrapper.find('#right-select').props().defaultValue).toBe('');
      });

      it('matches snapshot', () => {
        expect(wrapper).toMatchSnapshot();
      });

      it('renders JSONDiffViewer with latest version on both sides', () => {
        expect(wrapper.exists('JSONDiffViewer')).toBeTruthy();
        expect(wrapper.find('JSONDiffViewer').props().value).toStrictEqual([
          versions[0].schema,
          versions[0].schema,
        ]);
      });
    });

    describe('when schema versions are loaded and two versions in path', () => {
      const wrapper = shallow(
        setupWrapper({
          versions,
          leftVersionInPath: '1',
          rightVersionInPath: '2',
        })
      );

      it('renders left select with version 1', () => {
        expect(wrapper.find('#left-select')).toHaveLength(1);
        expect(wrapper.find('#left-select').props().defaultValue).toBe('1');
      });

      it('renders right select with version 2', () => {
        expect(wrapper.find('#right-select')).toHaveLength(1);
        expect(wrapper.find('#right-select').props().defaultValue).toBe('2');
      });

      it('matches snapshot', () => {
        expect(wrapper).toMatchSnapshot();
      });

      it('renders JSONDiffViewer with version 1 at left, version 2 at right', () => {
        expect(wrapper.exists('JSONDiffViewer')).toBeTruthy();
        expect(wrapper.find('JSONDiffViewer').props().value).toStrictEqual([
          JSON.stringify(JSON.parse(versions[2].schema), null, '\t'),
          JSON.stringify(JSON.parse(versions[1].schema), null, '\t'),
        ]);
      });
    });

    describe('when schema versions are loaded and only one versions in path', () => {
      const wrapper = shallow(
        setupWrapper({
          versions,
          leftVersionInPath: '1',
        })
      );

      it('renders left select with version 1', () => {
        expect(wrapper.find('#left-select')).toHaveLength(1);
        expect(wrapper.find('#left-select').props().defaultValue).toBe('1');
      });

      it('renders right select with empty value', () => {
        expect(wrapper.find('#right-select')).toHaveLength(1);
        expect(wrapper.find('#right-select').props().defaultValue).toBe('');
      });

      it('renders JSONDiffViewer with version 1 at left, version 3 at right', () => {
        expect(wrapper.exists('JSONDiffViewer')).toBeTruthy();
        expect(wrapper.find('JSONDiffViewer').props().value).toStrictEqual([
          JSON.stringify(JSON.parse(versions[2].schema), null, '\t'),
          versions[0].schema,
        ]);
      });

      it('matches snapshot', () => {
        expect(wrapper).toMatchSnapshot();
      });
    });
  });
});
