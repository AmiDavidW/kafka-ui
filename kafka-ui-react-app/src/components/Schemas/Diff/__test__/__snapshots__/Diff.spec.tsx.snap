// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Diff View when page with schema versions is loading matches snapshot 1`] = `
<div
  className="section"
>
  <PageLoader />
</div>
`;

exports[`Diff View when schema versions are loaded and no specified versions in path matches snapshot 1`] = `
<div
  className="section"
>
  <div
    className="box tile is-ancestor is-vertical"
  >
    <div
      className="tile"
    >
      <div
        className="tile is-6"
      >
        <div
          className="select"
        >
          <select
            defaultValue=""
            id="left-select"
            onChange={[Function]}
          >
            <option
              key="left-3"
              value="3"
            >
              Version 3
            </option>
            <option
              key="left-2"
              value="2"
            >
              Version 2
            </option>
            <option
              key="left-1"
              value="1"
            >
              Version 1
            </option>
          </select>
        </div>
      </div>
      <div
        className="tile is-6"
      >
        <div
          className="select"
        >
          <select
            defaultValue=""
            id="right-select"
            onChange={[Function]}
          >
            <option
              key="right-3"
              value="3"
            >
              Version 3
            </option>
            <option
              key="right-2"
              value="2"
            >
              Version 2
            </option>
            <option
              key="right-1"
              value="1"
            >
              Version 1
            </option>
          </select>
        </div>
      </div>
    </div>
    <styled.div>
      <DiffViewer
        isFixedHeight={false}
        schemaType="PROTOBUF"
        setOptions={
          Object {
            "autoScrollEditorIntoView": true,
          }
        }
        value={
          Array [
            "syntax = \\"proto3\\";
package com.indeed;

message MyRecord {
  int32 id = 1;
  string name = 2;
}
",
            "syntax = \\"proto3\\";
package com.indeed;

message MyRecord {
  int32 id = 1;
  string name = 2;
}
",
          ]
        }
      />
    </styled.div>
  </div>
</div>
`;

exports[`Diff View when schema versions are loaded and only one versions in path matches snapshot 1`] = `
<div
  className="section"
>
  <div
    className="box tile is-ancestor is-vertical"
  >
    <div
      className="tile"
    >
      <div
        className="tile is-6"
      >
        <div
          className="select"
        >
          <select
            defaultValue="1"
            id="left-select"
            onChange={[Function]}
          >
            <option
              key="left-3"
              value="3"
            >
              Version 3
            </option>
            <option
              key="left-2"
              value="2"
            >
              Version 2
            </option>
            <option
              key="left-1"
              value="1"
            >
              Version 1
            </option>
          </select>
        </div>
      </div>
      <div
        className="tile is-6"
      >
        <div
          className="select"
        >
          <select
            defaultValue=""
            id="right-select"
            onChange={[Function]}
          >
            <option
              key="right-3"
              value="3"
            >
              Version 3
            </option>
            <option
              key="right-2"
              value="2"
            >
              Version 2
            </option>
            <option
              key="right-1"
              value="1"
            >
              Version 1
            </option>
          </select>
        </div>
      </div>
    </div>
    <styled.div>
      <DiffViewer
        isFixedHeight={false}
        schemaType="PROTOBUF"
        setOptions={
          Object {
            "autoScrollEditorIntoView": true,
          }
        }
        value={
          Array [
            "{
	\\"type\\": \\"record\\",
	\\"name\\": \\"MyRecord1\\",
	\\"namespace\\": \\"com.mycompany\\",
	\\"fields\\": [
		{
			\\"name\\": \\"id\\",
			\\"type\\": \\"long\\"
		}
	]
}",
            "syntax = \\"proto3\\";
package com.indeed;

message MyRecord {
  int32 id = 1;
  string name = 2;
}
",
          ]
        }
      />
    </styled.div>
  </div>
</div>
`;

exports[`Diff View when schema versions are loaded and two versions in path matches snapshot 1`] = `
<div
  className="section"
>
  <div
    className="box tile is-ancestor is-vertical"
  >
    <div
      className="tile"
    >
      <div
        className="tile is-6"
      >
        <div
          className="select"
        >
          <select
            defaultValue="1"
            id="left-select"
            onChange={[Function]}
          >
            <option
              key="left-3"
              value="3"
            >
              Version 3
            </option>
            <option
              key="left-2"
              value="2"
            >
              Version 2
            </option>
            <option
              key="left-1"
              value="1"
            >
              Version 1
            </option>
          </select>
        </div>
      </div>
      <div
        className="tile is-6"
      >
        <div
          className="select"
        >
          <select
            defaultValue="2"
            id="right-select"
            onChange={[Function]}
          >
            <option
              key="right-3"
              value="3"
            >
              Version 3
            </option>
            <option
              key="right-2"
              value="2"
            >
              Version 2
            </option>
            <option
              key="right-1"
              value="1"
            >
              Version 1
            </option>
          </select>
        </div>
      </div>
    </div>
    <styled.div>
      <DiffViewer
        isFixedHeight={false}
        schemaType="PROTOBUF"
        setOptions={
          Object {
            "autoScrollEditorIntoView": true,
          }
        }
        value={
          Array [
            "{
	\\"type\\": \\"record\\",
	\\"name\\": \\"MyRecord1\\",
	\\"namespace\\": \\"com.mycompany\\",
	\\"fields\\": [
		{
			\\"name\\": \\"id\\",
			\\"type\\": \\"long\\"
		}
	]
}",
            "{
	\\"type\\": \\"record\\",
	\\"name\\": \\"MyRecord2\\",
	\\"namespace\\": \\"com.mycompany\\",
	\\"fields\\": [
		{
			\\"name\\": \\"id\\",
			\\"type\\": \\"long\\"
		}
	]
}",
          ]
        }
      />
    </styled.div>
  </div>
</div>
`;
